src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/ToAsyncEnumerable.cs

```diff
@@ -0,0 +1,70 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Linq
+{
+    public static partial class AsyncEnumerable
+    {
+        /// <summary>Creates a new <see cref="IAsyncEnumerable{T}"/> that iterates through <paramref name="source"/>.</summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">An <see cref="IEnumerable{T}"/> of the elements to enumerate.</param>
+        /// <returns>An <see cref="IAsyncEnumerable{T}"/> containing the sequence of elements from <paramref name="source"/>.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="source"/> is <see langword="null"/>.</exception>
+        /// <remarks>
+        /// Each iteration through the resulting <see cref="IAsyncEnumerable{T}"/> will iterate through the <paramref name="source"/>.
+        /// </remarks>
+        public static IAsyncEnumerable<TSource> ToAsyncEnumerable<TSource>(
+            this IEnumerable<TSource> source)
+        {
+            ThrowHelper.ThrowIfNull(source);
+
+            return source switch
+            {
+                TSource[] array => FromArray(array),
+                List<TSource> list => FromList(list),
+                IList<TSource> list => FromIList(list),
+                _ => FromIterator(source),
+            };
+
+            static async IAsyncEnumerable<TSource> FromArray(TSource[] source)
+            {
+                for (int i = 0; ; i++)
+                {
+                    TSource[] array = source;
```

Comment:
Can this be moved outside of the loop? Why is this so different than the other ones, is it performance related?

Reply:
If it were to be moved outside the loop, it'd become a field on the display class, just as `source` is, because it would span across a yield. And as a field, the JIT can't optimize away the bounds check on the subsequent array access, because the actual object in that field could have changed between the reads. As a local, the JIT can remove the bounds check. This is only relevant to arrays, because the others aren't checking purely against the array size, e.g. `List<T>` is actually comparing against a separate integer tracking the number of elements in its array, which is possibly larger than the number of elements contained.

Reply:
> In that case, i should also be stored in a variable, as currently, i is a field of the display class, which results in the fact that the JIT can't optimize away the bounds check.

Yes, that should be fixed.